src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ProvideStreamMemoryBasedAsyncOverrides.cs

```diff
@@ -0,0 +1,208 @@
+            }
+
+            var readAsyncArrayMethod = GetOverload(streamType, ReadAsyncName, byteArrayType, int32Type, int32Type, cancellationTokenType);
+            var readAsyncMemoryMethod = GetOverload(streamType, ReadAsyncName, memoryOfByteType, cancellationTokenType);
+            var writeAsyncArrayMethod = GetOverload(streamType, WriteAsyncName, byteArrayType, int32Type, int32Type, cancellationTokenType);
+            var writeAsyncMemoryMethod = GetOverload(streamType, WriteAsyncName, readOnlyMemoryOfByteType, cancellationTokenType);
+
+            if (readAsyncArrayMethod is null || readAsyncMemoryMethod is null || writeAsyncArrayMethod is null || writeAsyncMemoryMethod is null)
+            {
+                requiredSymbols = default;
+                return false;
+            }
+
+            requiredSymbols = new RequiredSymbols(
+                streamType, memoryOfByteType, readOnlyMemoryOfByteType,
+                readAsyncArrayMethod, readAsyncMemoryMethod,
+                writeAsyncArrayMethod, writeAsyncMemoryMethod);
+            return true;
+        }
+
+        private static IMethodSymbol? GetOverload(ITypeSymbol containingType, string methodName, params ITypeSymbol[] argumentTypes)
+        {
+            return containingType.GetMembers(methodName)
+                .SingleOrDefault(symbol => symbol is IMethodSymbol m && IsMatch(m, argumentTypes)) as IMethodSymbol;
+
+            static bool IsMatch(IMethodSymbol method, ITypeSymbol[] argumentTypes)
+            {
+                if (method.Parameters.Length != argumentTypes.Length)
+                    return false;
+
+                for (int index = 0; index < argumentTypes.Length; ++index)
+                {
+                    if (!argumentTypes[index].Equals(method.Parameters[index].Type, SymbolEqualityComparer.Default))
+                        return false;
+                }
+
+                return true;
+            }
+        }
+
+        //  If the specified type overrides the specified method on its immediate base class, returns the method symbol representing the 
+        //  overriding method. Returns null if the specified type does not override the specified method.
+        private static IMethodSymbol? GetSymbolForOverridingMethod(ITypeSymbol derivedType, IMethodSymbol overriddenMethod)
+        {
+            RoslynDebug.Assert(derivedType.BaseType.Equals(overriddenMethod.ContainingType, SymbolEqualityComparer.Default));
+
+            return derivedType.GetMembers(overriddenMethod.Name)
+                .SingleOrDefault(x =>
+                {
+                    return x.IsOverride && overriddenMethod.Equals(x.GetOverriddenMember(), SymbolEqualityComparer.Default);
+                }) as IMethodSymbol;
+        }
+
+        //  We will not be doing any comparisons on this type.
+#pragma warning disable CA1815 // Override equals and operator equals on value types
+        private readonly struct RequiredSymbols
```

Comment:
Not sure if we could consider a `record` here?

Reply:
I figured we could save on the allocation by using a struct, since it's not copied (it's literally just passed out of `TryGetRequiredSymbols` via an **out** parameter), and I've heard allocations are a pain point in the compiler.

Reply:
Fixed.